src/functions.erl

```diff
@@ -0,0 +1,124 @@
+-module(functions).
+
+-export([modified_functions/2,
+        callgraph/2]).
+
+-type fun_name()    :: string().
+-type line_num()    :: integer().
+-type boundaries()  :: {line_num(), line_num()}.
+-type ast()         :: erl_syntax:forms().
+-type tokens()      :: erl_scan:tokens().
+-type file_info()   :: {tokens(), ast()}. % TODO: This is a terrible name
+-type filename()    :: string().
+-type diffs()       :: [{filename(), {[line_num()], [line_num()]}}].
+
+% Finds all the functions in the AST, and pairs them up with their first and last line in the source
+-spec functions(tokens(), ast()) -> [{fun_name(), arity(), boundaries()}].
+functions(Tokens, AST) ->
+    Functions = lists:filter(fun(Tree) -> erl_syntax:type(Tree) =:= function end, AST),
+    StartLines = lists:map(fun(Tree) -> erl_syntax:get_pos(leftmost_node(Tree)) end, Functions),
+    RightmostLines = lists:map(fun(Tree) -> erl_syntax:get_pos(rightmost_node(Tree)) end, Functions),
+    EndLines = lists:map(fun(EndLine) -> offset(EndLine, Tokens) end, RightmostLines),
+    Names = lists:map(fun(Tree) -> erl_syntax:atom_value(erl_syntax:function_name(Tree)) end, Functions),
+    Arities = lists:map(fun erl_syntax:function_arity/1, Functions),
+    lists:zip3(Names, Arities, lists:zip(StartLines, EndLines)).
+
+
+-spec rightmost_node(ast()) -> ast().
+rightmost_node(Tree) ->
+    case erlang:is_list(Tree) of
+        true -> rightmost_node(lists:last(Tree));
+        false ->
+            Children = erl_syntax:subtrees(Tree),
+            case lists:reverse(Children) of
+                [] -> Tree;
+                [RightmostChild | _] ->
+                    case RightmostChild of
+                        [] -> Tree;
+                        _  -> rightmost_node(RightmostChild)
+                    end
+            end
+    end.
+
+-spec leftmost_node(ast()) -> ast().
+leftmost_node(Tree) ->
+    case erlang:is_list(Tree) of
+        true -> leftmost_node(hd(Tree));
+        false ->
+            Children = erl_syntax:subtrees(Tree),
+            case Children of
+                [] -> Tree;
+                [LeftMostChild | _] -> leftmost_node(LeftMostChild)
+            end
+    end.
+
+% Finds the number of lines after the last meaningful node in the AST
+% This is needed because we want the last line of the function based on the
+% source file, not from the AST, which doesn't deal with dots and other "unnecessary"
+% syntactic elements
+-spec offset(line_num(), tokens()) -> line_num().
+offset(LineNum, Tokens) ->
+    Rest = lists:dropwhile(fun(T) -> element(2, T) =/= LineNum end, Tokens),
```

Comment:
@daniel-horpacsi  I'm using element here, because the token could be either a 2 or a 3-tuple.
https://www.erlang.org/doc/man/erl_scan.html#type-token

Reply:
I've looked at this again, but to use pattern matching, I would have to replace each call to `element` with its own named function, each having two cases. Or am I just not getting something?

Reply:
I mean I would be totally okay with extracting it into a separate function, but I guess 3 would be a bit overkill.

Reply:
Nevermind. Should have looked into erl_scan in the first place.